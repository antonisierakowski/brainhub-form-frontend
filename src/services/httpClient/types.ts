export enum StatusCode {
  OK = 200,
  UNPROCESSABLE_ENTITY = 422,
  INTERNAL_ERROR = 500,
}

export interface ApiResponse<TBody = any> {
  status: StatusCode;
  response: TBody;
}
