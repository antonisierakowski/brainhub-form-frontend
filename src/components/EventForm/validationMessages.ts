import { MAX_NAME_LENGTH, MIN_NAME_LENGTH } from '../../constants';

export const createNameTooShortMessage = (humanReadableFieldName: string) => `${humanReadableFieldName} should be at least ${MIN_NAME_LENGTH} characters long`;
export const createNameTooLongMessage = (humanReadableFieldName: string) => `${humanReadableFieldName} should be shorten than ${MAX_NAME_LENGTH} characters`;

export const fieldIsRequiredMessage = 'This field is required';

export const validEmailMessage = 'Please provide a valid e-mail address';
export const validDateMessage = 'Please provide a valid date';
