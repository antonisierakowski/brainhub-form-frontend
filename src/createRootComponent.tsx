import React from 'react';
import { Provider } from 'react-redux';
import { store } from './store';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';
import { Store } from 'redux';

export const createRootComponent = (
  component: React.ReactElement,
  reduxStore: Store = store,
): React.ReactElement => (
  <Provider store={reduxStore}>
    <MuiPickersUtilsProvider utils={MomentUtils}>
      {component}
    </MuiPickersUtilsProvider>
  </Provider>
);
