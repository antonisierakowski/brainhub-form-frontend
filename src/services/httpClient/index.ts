import { BACKEND_DOMAIN } from '../../constants';
import * as Axios from 'axios';
import * as exceptions from './exceptions';
import { EventModel } from '../../store/request/model';
import { ApiResponse, StatusCode } from './types';

export const axiosInstance = Axios.default.create({
  baseURL: BACKEND_DOMAIN,
});

export class HttpClient {
  private static instance: HttpClient;
  private axiosInstance: Axios.AxiosInstance;

  private constructor() {
    this.axiosInstance = axiosInstance;
  }

  static getInstance(): HttpClient {
    if (!HttpClient.instance) {
      HttpClient.instance = new HttpClient();
    }

    return HttpClient.instance;
  }

  private throwBasedOnStatusCode(response: Axios.AxiosResponse) {
    switch (response.status) {
      case StatusCode.OK: {
        return response;
      }
      case StatusCode.UNPROCESSABLE_ENTITY: {
        throw new exceptions.ApiValidationError();
      }
      case StatusCode.INTERNAL_ERROR:
      default: {
        throw new exceptions.ApiInternalError();
      }
    }
  }

  private async post<TBody = any, TResponse = ApiResponse>(
    url: string, body?: TBody,
  ): Promise<TResponse> {
    try {
      const response = await this.axiosInstance.post(url, body);
      return response.data;
    } catch(error) {
      if (!error.response) {
        throw new exceptions.NoApiResponseError();
      }
      this.throwBasedOnStatusCode(error.response);
    }
  }

  submitEvent = async (event: EventModel): Promise<ApiResponse> => {
    const endpoint = '/event';
    return await this.post<EventModel>(endpoint, event);
  }

}

export default HttpClient.getInstance();
