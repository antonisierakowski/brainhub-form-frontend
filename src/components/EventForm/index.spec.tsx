import React from 'react';
import { EventForm } from './index';
import { render, RenderResult, fireEvent, wait } from '@testing-library/react';
import { createRootComponent } from '../../createRootComponent';
import { act } from 'react-dom/test-utils';
import { fieldIsRequiredMessage } from './validationMessages';
import {
  emailInvalid,
  nameTooLong,
  nameTooShort,
  validEventFixture,
} from './validationSchema.spec';
import httpClient from '../../services/httpClient/index';

let renderResult: RenderResult;

beforeEach(() => {
  renderResult = render(createRootComponent(<EventForm />));
});

describe('EventForm component', () => {
  it('should match to generated snapshot', () => {
    expect(renderResult.asFragment).toMatchSnapshot();
  });

  it('should validate all fields on submit click (with date being initially valid)', async () => {
    const submitButton = renderResult.getByTestId('submitButton');
    fireEvent.click(submitButton);
    await wait(() => {
      expect(renderResult.queryAllByText(fieldIsRequiredMessage)).toHaveLength(
        3,
      );
    });
  });

  it('should validate firstName field on blur', async () => {
    const firstNameInput = renderResult.container.querySelector(
      'input[name="firstName"]',
    );
    act(() => {
      fireEvent.input(firstNameInput, { target: { value: nameTooShort } });
      fireEvent.blur(firstNameInput);
    });
    await wait(() => {
      expect(renderResult.queryAllByText(fieldIsRequiredMessage)).toHaveLength(
        1,
      );
    });
  });

  it('should validate lastName field on blur', async () => {
    const lastNameInput = renderResult.container.querySelector(
      'input[name="lastName"]',
    );
    act(() => {
      fireEvent.input(lastNameInput, { target: { value: nameTooLong } });
      fireEvent.blur(lastNameInput);
    });
    await wait(() => {
      expect(renderResult.queryAllByText(fieldIsRequiredMessage)).toHaveLength(
        1,
      );
    });
  });

  it('should validate email field on blur', async () => {
    const emailInput = renderResult.container.querySelector(
      'input[name="email"]',
    );
    act(() => {
      fireEvent.input(emailInput, { target: { value: emailInvalid } });
      fireEvent.blur(emailInput);
    });
    await wait(() => {
      expect(renderResult.queryAllByText(fieldIsRequiredMessage)).toHaveLength(
        1,
      );
    });
  });

  it('should call httpClient method', async () => {
    const mockHttpCall = jest.fn();
    jest.spyOn(httpClient, 'submitEvent').mockImplementation(mockHttpCall);

    const firstNameInput = renderResult.container.querySelector(
      'input[name="firstName"]',
    );
    const lastNameInput = renderResult.container.querySelector(
      'input[name="lastName"]',
    );
    const emailInput = renderResult.container.querySelector(
      'input[name="email"]',
    );
    const dateInput = renderResult.container.querySelector(
      'input[name="date"]',
    );
    const submitButton = renderResult.getByTestId('submitButton');
    act(() => {
      fireEvent.input(firstNameInput, {
        target: { value: validEventFixture.firstName },
      });
      fireEvent.blur(firstNameInput);
      fireEvent.input(lastNameInput, {
        target: { value: validEventFixture.lastName },
      });
      fireEvent.blur(lastNameInput);
      fireEvent.input(emailInput, {
        target: { value: validEventFixture.email },
      });
      fireEvent.blur(emailInput);
      fireEvent.input(dateInput, { target: { value: validEventFixture.date } });
      fireEvent.blur(dateInput);
    });

    act(() => {
      fireEvent.click(submitButton);
    });

    expect(renderResult.queryAllByText(fieldIsRequiredMessage)).toHaveLength(0);

    await wait(() => {
      expect(mockHttpCall).toHaveBeenCalledTimes(1);
    });
  });
});
