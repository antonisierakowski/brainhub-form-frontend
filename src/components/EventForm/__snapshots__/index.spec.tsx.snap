// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm component should match to generated snapshot 1`] = `[Function]`;
